### PR TITLE
lightning/parser: support STARTING BY

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/40499

close https://github.com/pingcap/tidb/pull/40801

Problem Summary:

as title

### What is changed and how it works?

- STARTING BY: read the line by `ReadUntilTerminator` and check the content. If it contains startingBy, strip the prefix and put back to parser buffer.

Also check that caller can use `ReadUntilTerminator` to skip N lines as IGNORE N LINES

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/mydump/csv_parser.go [hunk 1]
```diff
@@ -370,11 +378,43 @@ func (parser *CSVParser) readRecord(dst []string) ([]string, error) {
...
 outside:
 	for {
+		// we should drop
+		// 1. the whole line if it does not contain startingBy
+		// 2. any character before startingBy
+		// since we have checked startingBy does not contain terminator, we can
+		// split at terminator to check the substring contains startingBy. Even
+		// if the terminator is inside a quoted field which means it's not the
+		// end of a line, the substring can still be dropped by rule 2.
+		if len(parser.startingBy) > 0 && !foundStartingByThisLine {
```
Comment (gozssky) [inline br/pkg/lightning/mydump/csv_parser.go:394]:
```suggestion
		for len(parser.startingBy) > 0 {
```

File: br/pkg/lightning/mydump/csv_parser.go [hunk 2]
```diff
@@ -370,11 +378,43 @@ func (parser *CSVParser) readRecord(dst []string) ([]string, error) {
...
+			}
+			idx := bytes.Index(content, parser.startingBy)
+			if idx == -1 {
+				continue
+			}
+			foundStartingByThisLine = true
+			content = content[idx+len(parser.startingBy):]
+			content = append(content, parser.newLine...)
+			parser.buf = append(content, parser.buf...)
+			parser.pos = oldPos + int64(idx+len(parser.startingBy))
```
Comment (gozssky) [inline br/pkg/lightning/mydump/csv_parser.go:415]:
```suggestion
			parser.pos = oldPos + int64(idx+len(parser.startingBy))
			break
```

File: br/pkg/lightning/config/config.go [hunk 3]
```diff
@@ -455,6 +455,8 @@ type CSVConfig struct {
 	TrimLastSep     bool   `toml:"trim-last-separator" json:"trim-last-separator"`
 	NotNull         bool   `toml:"not-null" json:"not-null"`
 	BackslashEscape bool   `toml:"backslash-escape" json:"backslash-escape"`
+	// hide these options for lightning configuration file, they can only be used by LOAD DATA
+	StartingBy string `toml:"-" json:"-"`
```
Comment (gozssky) [inline br/pkg/lightning/config/config.go:459]: Please add a doc link for `StartingBy`.

File: br/pkg/lightning/mydump/csv_parser.go (hunk 1 above)
Comment (gozssky) [reply br/pkg/lightning/mydump/csv_parser.go:394]: Nit: maybe more readable

Comment (lance6716) [reply br/pkg/lightning/mydump/csv_parser.go:394]: when the `outside` for loop goes to next round, it does not mean a line is processed (it only means we have met a special token). So we need another variable to record the state per line.

File: br/pkg/lightning/mydump/csv_parser_test.go [hunk 4]
```diff
@@ -935,6 +964,211 @@ func TestTerminator(t *testing.T) {
...
+			Delimiter:  `"`,
+			Terminator: "\n",
+			StartingBy: "x\nxx",
+		},
+	}
+	_, err := mydump.NewCSVParser(context.Background(), &cfg.CSV, nil, 1, ioWorkers, false, nil)
+	require.ErrorContains(t, err, "starting-by cannot contain (line) terminator")
+}
+
+func TestCallerCanIgnoreNLines(t *testing.T) {
```
Comment (gozssky) [inline br/pkg/lightning/mydump/csv_parser_test.go:1098]: It seems it's better to move this test to `LOAD DATA`. The parser shouldn't care about how the caller uses it.

**Review by gozssky — COMMENTED** [review]
Rest LGTM.

File: br/pkg/lightning/mydump/csv_parser_test.go (hunk 4 above)
Comment (lance6716) [reply br/pkg/lightning/mydump/csv_parser_test.go:1098]: OK, I'll do it in next PR

**Review by gozssky — APPROVED** [review]

**Comment by lance6716** [comment]
ping @lichunzhu @GMHDBJD @dsdashun

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**